### PR TITLE
topology2: set DMIC0 pcm id to 6

### PR DESCRIPTION
**Pull request by libinyang**
UCM requires DMIC PCM ID to be 6.
Fixes: #4068
Signed-off-by: Libin Yang <libin.yang@intel.com>

### CODE REVIEW

**Review by plbossart — CHANGES_REQUESTED** [review]

File: tools/topology/topology2/platform/intel/dmic-generic.conf [hunk 1]
```diff
@@ -124,9 +124,9 @@ Object.Base.route.100 {
 }
 
 Object.PCM {
-	pcm.10 {
+	pcm.6 {
 		name	"DMIC"
-		id 10
+		id 6
```
[the remark this answers is not in the document]
Comment (libinyang) [reply tools/topology/topology2/platform/intel/dmic-generic.conf:129]: @plbossart Thanks for the suggestion. I will refine the patch by using a parameter based on the UCM for each mode.

**Comment by libinyang** [comment]
v2 update:
use DMIC0_PCM_ID for dmic0 pcm id setting.

**Comment by lgirdwood** [comment]
> This PR depends on #6787. Without #6787, sof hda will fail to create the snd card because of the conflict between HDMI PCM ID and DMIC0 PCM ID, which are both 6.

#6787 now pending final approval from @ranj063

**Comment by mengdonglin** [comment]
@plbossart @RanderWang Can you please check if this PR can be merged? It depends on https://github.com/thesofproject/sof/pull/6787 that has been merged.

**Comment by libinyang** [comment]
https://github.com/thesofproject/sof/pull/6787 is merged, but there are some more discussions on in https://github.com/thesofproject/sof/pull/6953. Let's wait for the final conclusion.

**Comment by lgirdwood** [comment]
> #6787 is merged, but there are some more discussions in #6953. Let's wait for the final conclusion.

#6953 closed. @libinyang @bardliao @kv2019i whats the next move now ?

**Review by bardliao — APPROVED** [review]

**Review by kv2019i — APPROVED** [review]
Looks good. There's a conflict now though, so needs an update.

**Comment by libinyang** [comment]
Patches update

**Review by lgirdwood — APPROVED** [review]
@plbossart changes good now ?

**Comment by ranj063** [comment]
> sounds incomplete to me. I am pretty sure the device 10 is not right for cavs-sdw topologies

@plbossart this may be incomplete but it doesnt make the existing topologies any worse right? We're currently setting 10 for all PCM IDs for all topologies. At least the HDA ones are fixed with this one

**Comment by plbossart** [comment]
> > sounds incomplete to me. I am pretty sure the device 10 is not right for cavs-sdw topologies
> 
> @plbossart this may be incomplete but it doesnt make the existing topologies any worse right? We're currently setting 10 for all PCM IDs for all topologies. At least the HDA ones are fixed with this one

what the point of doing things one step at a time, apart from wasting reviewer time?
This is a shared module, so we do things right. I don't want to review this multiple times.

**Comment by libinyang** [comment]
> what the point of doing things one step at a time, apart from wasting reviewer time? This is a shared module, so we do things right. I don't want to review this multiple times.
> 
> Edit: I provided the SAME feedback last month in [#6904 (comment)](https://github.com/thesofproject/sof/pull/6904#discussion_r1060760840). Seriously?

@plbossart We have a rule for the DMIC0 PCM ID described in https://github.com/thesofproject/sof/pull/6783. The DMIC0 PCM ID should be 10.
For the below topologies, I checked: what PCMID is set in topology1, what UCM requires the DMIC0 PCMID to be.
1. cavs-nocodec-bt.conf:
    There seems no counterpart in topology1 and no special requirement from UCM.
2. cavs-nocodec.conf 
    The PCMID is 10 in topology1 and no special requirement from UCM.
3. cavs-rt5682.conf 
    The PCMID is 99 in topology1 (We should not use 99 anymore, I think) and no special requirement from UCM.
4. cavs-sdw.conf 
    The PCMID is 10 and UCM2 can accept any PCMID for sof-soundwire card (no special requirement from UCM).
5. sof-hda-generic.conf. 
    The PCMID is 6 and UCM2 requires PCMID to be 6. 

So I think this PR can cover all the scenarios in topology2: set DMIC0 PCMID to 6 only on HDA platform and let other platforms to be 10.

**Comment by plbossart** [comment]
> 3. cavs-rt5682.conf
>    The PCMID is 99 in topology1 (We should not use 99 anymore, I think) and no special requirement from UCM.

have you asked the Google/Chrome folks? They use UCM, don't they? @cujomalainey @sathya-nujella 
It's not our decision to change the device numbers, and there is nothing in the card name that identifies MTL/topology2/

> 4. cavs-sdw.conf
>    The PCMID is 10 and UCM2 can accept any PCMID for sof-soundwire card (no special requirement from UCM).

the requirement is this:
CapturePCM "hw:${CardId},${find-device:type=pcm,stream=capture,field=id,regex='DMIC '}"

is this met?

**Comment by plbossart** [comment]
> > > 3. cavs-rt5682.conf
> > >    The PCMID is 99 in topology1 (We should not use 99 anymore, I think) and no special requirement from UCM.
> > 
> > 
> > have you asked the Google/Chrome folks? They use UCM, don't they? @cujomalainey @sathya-nujella It's not our decision to change the device numbers, and there is nothing in the card name that identifies MTL/topology2/
> 
> Topology1 based Chrome systems are using 99 for DMIC capture. Can you please limit the change to topology2 only if possible? If it is affecting topology2 only, then from MTL onwards we can update UCM's in Chromium gits accordingly.

if we change only for topology2, you won't be able to test this topology on existing TGL chromebooks. I see no value in changing the device number and make integration/testing more complicated. Yes 99 is not a very good number, but it's not worse than 10. It's easier to stick with legacy.

**Comment by libinyang** [comment]
> > 
> > Topology1 based Chrome systems are using 99 for DMIC capture. Can you please limit the change to topology2 only if possible? If it is affecting topology2 only, then from MTL onwards we can update UCM's in Chromium gits accordingly.
> 
> if we change only for topology2, you won't be able to test this topology on existing TGL chromebooks. I see no value in changing the device number and make integration/testing more complicated. Yes 99 is not a very good number, but it's not worse than 10. It's easier to stick with legacy.

@plbossart @sathya-nujella Thanks for the feedback. I will follow @plbossart 's suggestion and make the DMIC0 PCMID to be 99 for sof-mtl-max98357a-rt5682 platforms on topology2, and topology1 will not be impacted. BTW: Our current rt5682 in tplg2 is used for sof-mtl-max98357a-rt5682 platforms. However, rt5682 can also be used in other combinations, for example, in tplg1, it can be used in sof-cml-rt1011-rt5682 and sof-cml-rt5682-max98357a, and the dmic0 PCMID is 1 on such platforms. To make it more flexible, I will set the ```DMIC0_PCM_ID=99``` in the CMakefile.

> the requirement is this:
> CapturePCM "hw:${CardId},${find-device:type=pcm,stream=capture,field=id,regex='DMIC '}"

> is this met?

Currently, it meets the requirement on our supported tplg2 platforms. But this may be conflicted with DMIC1, for example WoV PCM name. I will have a check and test. If there is conflict, I will fix it in another PR as this is a PCM name issue, not PCM ID.

**Comment by libinyang** [comment]
patch updated:
1. set sof-mtl-max98357a-rt5682 DMIC0 pcm id to 99

**Review by plbossart — APPROVED** [review]

**Comment by mengdonglin** [comment]
@plbossart @kv2019i @lgirdwood Are we good to merge this PR?

**Comment by ranj063** [comment]
> @plbossart @kv2019i @lgirdwood Are we good to merge this PR?

@libinyang @mengdonglin there are conflicts

**Comment by libinyang** [comment]
Patches are updated based on the latest code.

**Pull request merged**